File: original/Code/Combat/transitiongameobj.cpp
#include "global.h"
#include "transitiongameobj.h"
#include "persistfactory.h"
#include "combatchunkid.h"
#include "simpledefinitionfactory.h"
#include "combat.h"

/*
** TransitionGameObjDef
*/

SimplePersistFactoryClass<TransitionGameObjDef, CHUNKID_GAME_OBJECT_DEF_TRANSITION>	_TransitionGameObjDefPersistFactory;

DECLARE_DEFINITION_FACTORY(TransitionGameObjDef, CLASSID_GAME_OBJECT_DEF_TRANSITION, "Transition") _TransitionGameObjDefDefFactory;

TransitionGameObjDef::~TransitionGameObjDef( void )
{
	Free_Transition_List();
}

uint32_t	TransitionGameObjDef::Get_Class_ID (void) const	
{ 
	return CLASSID_GAME_OBJECT_DEF_TRANSITION; 
}

PersistClass *	TransitionGameObjDef::Create( void ) const 
{
	TransitionGameObj * obj = new TransitionGameObj;
	obj->Init( *this );
	return obj;
}

enum	{
	CHUNKID_DEF_PARENT							= 1111991201,
	CHUNKID_DEF_TRANSITION,
};

bool	TransitionGameObjDef::Save( ChunkSaveClass & csave )
{
	csave.Begin_Chunk( CHUNKID_DEF_PARENT );
		BaseGameObjDef::Save( csave );
	csave.End_Chunk();

	/*
	**	Save each of the transition 'definitions' in our list.
	*/
	for( int index = 0; index < Transitions.Count (); index ++ ) {
		TransitionDataClass *transition = Transitions[index];
		if( transition != NULL ) {

			/*
			**	Save this transition 'defintion' to its own chunk.
			*/
			csave.Begin_Chunk( CHUNKID_DEF_TRANSITION );
			transition->Save( csave );
			csave.End_Chunk();
		}
	}

	return true;
}

bool	TransitionGameObjDef::Load( ChunkLoadClass &cload )
{
	Free_Transition_List ();

	while (cload.Open_Chunk()) {
		switch(cload.Cur_Chunk_ID()) {

			case CHUNKID_DEF_PARENT:
				BaseGameObjDef::Load( cload );
				break;

			case CHUNKID_DEF_TRANSITION:
			{
				TransitionDataClass *transition = new TransitionDataClass;
				transition->Load( cload );
				Transitions.Add( transition );
			}
			break;
								
			default:
				Debug_Say(( "Unrecognized TransitionDef chunkID\n" ));
				break;

		}
		cload.Close_Chunk();
	}

	return true;
}

/*
**
*/
void TransitionGameObjDef::Free_Transition_List( void )
{
	/*
	**	Delete each of the transition 'definitions' in our list.
	*/
	for( int index = 0; index < Transitions.Count (); index ++ ) {
		TransitionDataClass *transition = Transitions[index];
		if( transition != NULL ) {
			delete transition;
		}
	}
	
	Transitions.Delete_All();
	return ;
}

const PersistFactoryClass & TransitionGameObjDef::Get_Factory (void) const 
{ 
	return _TransitionGameObjDefPersistFactory; 
}

/*
** TransitionGameObj
*/
SimplePersistFactoryClass<TransitionGameObj, CHUNKID_GAME_OBJECT_TRANSITION>	_TransitionGameObjPersistFactory;

const PersistFactoryClass & TransitionGameObj::Get_Factory (void) const 
{
	return _TransitionGameObjPersistFactory;
}

TransitionGameObj::TransitionGameObj()	:
	LadderIndex( -1 )
{
}

TransitionGameObj::~TransitionGameObj()
{
	Destroy_Transitions();
}

/*
**
*/
void TransitionGameObj::Init( void )
{
	Init( Get_Definition() );
}

void	TransitionGameObj::Init( const TransitionGameObjDef & definition )
{
	BaseGameObj::Init( definition );

	Create_Transitions();
}

const TransitionGameObjDef & TransitionGameObj::Get_Definition( void ) const
{
	return (const TransitionGameObjDef &)BaseGameObj::Get_Definition();
}

/*
** TransitionGameObj Save and Load
*/
enum	{
	CHUNKID_PARENT							=	1111991206,
	CHUNKID_VARIABLES,

	MICROCHUNKID_TRANSFORM				= 1,
	MICROCHUNKID_LADDER_INDEX
};

bool	TransitionGameObj::Save( ChunkSaveClass & csave )
{
	csave.Begin_Chunk( CHUNKID_PARENT );
	BaseGameObj::Save( csave );
	csave.End_Chunk();

	csave.Begin_Chunk( CHUNKID_VARIABLES );
	WRITE_MICRO_CHUNK( csave, MICROCHUNKID_TRANSFORM,		TM );
	WRITE_MICRO_CHUNK( csave, MICROCHUNKID_LADDER_INDEX,	LadderIndex );
	csave.End_Chunk();

	return true;
}

bool	TransitionGameObj::Load( ChunkLoadClass &cload )
{
	while (cload.Open_Chunk()) {
		switch(cload.Cur_Chunk_ID()) {

			case CHUNKID_PARENT:
				BaseGameObj::Load( cload );
				break;

			case CHUNKID_VARIABLES:
			{
				while (cload.Open_Micro_Chunk()) {
					switch(cload.Cur_Micro_Chunk_ID()) {
						READ_MICRO_CHUNK( cload, MICROCHUNKID_TRANSFORM,		TM );
						READ_MICRO_CHUNK( cload, MICROCHUNKID_LADDER_INDEX,	LadderIndex );
						default:
							Debug_Say(( "Unrecognized TransitionGameObj Variable chunkID\n" ));
							break;
					}
					cload.Close_Micro_Chunk();
				}

				break;
			}

			default:
				Debug_Say(( "Unrecognized Transition chunkID\n" ));
				break;

		}
		cload.Close_Chunk();
	}
	
	SaveLoadSystemClass::Register_Post_Load_Callback(this);
	return true;
}

void	TransitionGameObj::On_Post_Load( void )
{
	BaseGameObj::On_Post_Load();
	Create_Transitions();
	return ;
}

void	TransitionGameObj::Set_Transform(const Matrix3D & tm)
{
	TM = tm;
}

const Matrix3D &	TransitionGameObj::Get_Transform(void) const
{
	return TM;
}

void	TransitionGameObj::Create_Transitions( void )
{
	Destroy_Transitions();

	// Only create the transitions if they haven't already been created
	if( TransitionInstances.Count() == 0 ) {

		const TRANSITION_DATA_LIST & trans_data_list = Get_Definition().Get_Transition_List();
		Matrix3D tm = Get_Transform();

		for ( int i = 0; i < trans_data_list.Count(); i++ ) {
			// make new instance
			TransitionInstanceClass * trans = new TransitionInstanceClass( *trans_data_list[i] );
			// setup
			trans->Set_Parent_Transform( tm );
			// add to our list
			TransitionInstances.Add( trans );
			// add to master list
			TransitionManager::Add( trans );
		}
	}

	//
	//	Ensure all the transition instances know what ladder they belong to (may
	// not apply for other transitions other than ladder).
	//
	Set_Ladder_Index( LadderIndex );
	return ;
}

void	TransitionGameObj::Destroy_Transitions( void )
{
	while ( TransitionInstances.Count() ) {
  		TransitionManager::Destroy( TransitionInstances[0] );
  		TransitionInstances.Delete( 0 );
	}

	// They get put in a list, so flush the list.
	TransitionManager::Destroy_Pending();
}

void	TransitionGameObj::Update_Transitions( void )
{
	Matrix3D tm = Get_Transform();

	// Update all transitions
	for ( int i = 0; i < TransitionInstances.Count(); i++ ) {
		TransitionInstances[i]->Set_Parent_Transform( tm );
		TransitionInstances[i]->Set_Ladder_Index( LadderIndex );
	}
}

void	TransitionGameObj::Set_Ladder_Index( int ladder_index )
{
	LadderIndex = ladder_index;

	// Update all transitions
	for ( int i = 0; i < TransitionInstances.Count(); i++ ) {
		TransitionInstances[i]->Set_Ladder_Index( LadderIndex );
	}

	return ;
}

File: kotlin-server/server/src/main/kotlin/ccr/server/net/TransitionGameObj.kt
package ccr.server.net

import ccr.math.Matrix3D
import ccr.server.defs.TransitionGameObjDef

// C++: TransitionGameObj : public BaseGameObj
// Hierarchy: NetworkObject → BaseGameObj → TransitionGameObj
class TransitionGameObj : BaseGameObj() {

    // C++: DynamicVectorClass<TransitionInstanceClass*> TransitionInstances
    var transitionInstances: MutableList<TransitionInstanceClass> = mutableListOf()

    // C++: Matrix3D TM (initialized to identity)
    var tm: Matrix3D = Matrix3D()

    // C++: int LadderIndex (initialized to -1 via constructor initializer list)
    // @JvmName avoids clash with fun getLadderIndex() / fun setLadderIndex() below
    @get:JvmName("ladderIndexField") @set:JvmName("setLadderIndexField")
    var ladderIndex: Int = -1

    // C++: ~TransitionGameObj() — calls Destroy_Transitions()
    override fun destruct() {
        destroyTransitions()
        super.destruct()
    }

    // C++: virtual void Init()
    override fun init() {
        init(getDefinition())
    }

    // C++: void Init(const TransitionGameObjDef & definition)
    fun init(definition: TransitionGameObjDef) {
        super.init(definition)
        createTransitions()
    }

    // C++: const TransitionGameObjDef & Get_Definition() const
    fun getDefinition(): TransitionGameObjDef = definition as TransitionGameObjDef

    // C++: virtual bool Save(ChunkSaveClass&)
    override fun save(csave: ChunkSaveClass): Boolean {
        csave.beginChunk(CHUNKID_PARENT)
        super.save(csave)
        csave.endChunk()

        csave.beginChunk(CHUNKID_VARIABLES)
        csave.writeMicroChunk(MICROCHUNKID_TRANSFORM,    tm)
        csave.writeMicroChunk(MICROCHUNKID_LADDER_INDEX, ladderIndex)
        csave.endChunk()

        return true
    }

    // C++: virtual bool Load(ChunkLoadClass&)
    override fun load(cload: ChunkLoadClass): Boolean {
        while (cload.openChunk()) {
            when (cload.curChunkId) {
                CHUNKID_PARENT -> super.load(cload)
                CHUNKID_VARIABLES -> {
                    while (cload.openMicroChunk()) {
                        when (cload.curMicroChunkId) {
                            MICROCHUNKID_TRANSFORM    -> tm          = cload.readMatrix3D()
                            MICROCHUNKID_LADDER_INDEX -> ladderIndex = cload.readInt()
                            else -> error("Unrecognized TransitionGameObj variable chunk ID: ${cload.curMicroChunkId}")
                        }
                        cload.closeMicroChunk()
                    }
                }
                else -> error("Unrecognized TransitionGameObj chunk ID: ${cload.curChunkId}")
            }
            cload.closeChunk()
        }

        // FIXME: SaveLoadSystemClass::Register_Post_Load_Callback(this) — pointer remap not ported
        return true
    }

    // C++: virtual void On_Post_Load()
    override fun onPostLoad() {
        super.onPostLoad()
        createTransitions()
    }

    // C++: void Set_Transform(const Matrix3D & tm)
    fun setTransform(tm: Matrix3D) {
        this.tm = tm
    }

    // C++: const Matrix3D & Get_Transform() const
    fun getTransform(): Matrix3D = tm

    // C++: void Create_Transitions()
    fun createTransitions() {
        destroyTransitions()

        // Only create the transitions if they haven't already been created
        check(transitionInstances.isEmpty())
        if (transitionInstances.isEmpty()) {
            val transDataList = getDefinition().transitions
            val currentTm = getTransform()

            for (transData in transDataList) {
                // make new instance
                val trans = TransitionInstanceClass(transData)
                // setup
                trans.setParentTransform(currentTm)
                // add to our list
                transitionInstances.add(trans)
                // add to master list
                TransitionManager.add(trans)
            }
        }

        // Ensure all the transition instances know what ladder they belong to
        setLadderIndex(ladderIndex)
    }

    // C++: void Destroy_Transitions()
    fun destroyTransitions() {
        while (transitionInstances.isNotEmpty()) {
            TransitionManager.destroy(transitionInstances[0])
            transitionInstances.removeAt(0)
        }

        // They get put in a list, so flush the list.
        TransitionManager.destroyPending()
    }

    // C++: void Update_Transitions()
    fun updateTransitions() {
        val currentTm = getTransform()

        // Update all transitions
        for (trans in transitionInstances) {
            trans.setParentTransform(currentTm)
            trans.setLadderIndex(ladderIndex)
        }
    }

    // C++: int Get_Transition_Count() const
    fun getTransitionCount(): Int = transitionInstances.size

    // C++: TransitionInstanceClass* Get_Transition(int index)
    fun getTransition(index: Int): TransitionInstanceClass = transitionInstances[index]

    // C++: int Get_Ladder_Index()
    fun getLadderIndex(): Int = ladderIndex

    // C++: void Set_Ladder_Index(int ladder_index)
    fun setLadderIndex(ladderIndex: Int) {
        this.ladderIndex = ladderIndex

        // Update all transitions
        for (trans in transitionInstances) {
            trans.setLadderIndex(ladderIndex)
        }
    }

    companion object {
        // Chunk IDs from transitiongameobj.cpp (TransitionGameObj)
        private const val CHUNKID_PARENT    = 1111991206
        private const val CHUNKID_VARIABLES = 1111991207

        // Micro chunk IDs from transitiongameobj.cpp
        private const val MICROCHUNKID_TRANSFORM    = 1
        private const val MICROCHUNKID_LADDER_INDEX = 2
    }
}
